device/modbus_mapper/src/devicetwin.js

```diff
@@ -0,0 +1,230 @@
+                    common.switchByte(internalData, (switchedData)=>{
+                        callback(null, switchedData);
+                    });
+                } else {
+                    callback(null, internalData);
+                }
+            }
+        ], function(err, transedData) {
+            transData = transedData;
+        });
+        this.transferDataType(visitor, property, transData, callback);
+    }
+
+    // transferDataType transfer data according to the dpl configuration
+    transferDataType(visitor, property, data, callback) {
+        let transData;
+        switch(property.datatype) {
+            case 'int':
+            case 'float':
+                if (visitor.visitorConfig.register === 'DiscreteInputRegister' || visitor.visitorConfig.register === 'CoilRegister') {
+                    common.bitArrayToInt(data, (num)=>{
+                        transData = num;
+                    });
+                } else if (visitor.visitorConfig.register === 'HoldingRegister' || visitor.visitorConfig.register === 'InputRegister') {
+                    common.twoByteArrayToInt(data, (num)=>{
+                        transData = num;
+                    })
+                }
+
+                if (visitor.visitorConfig.scale !=0 && transData != null) {
+                    transData = transData * visitor.visitorConfig.scale;
+                }
+
+                if (property.datatype === 'int') {
+                    transData = parseInt(transData);
+                }
+
+                if (property.maximum !== null && transData > parseFloat(property.maximum)) {
+                    logger.info("read data is larger than max value, use max value")
+                    transData = parseInt(property.maximum);
+                } else if (property.minimum !== null && transData < parseFloat(property.minimum)) {
+                    logger.info("read data is smaller than min value, use min value")
+                    transData = parseInt(property.minimum);
+                }
+
+                callback(transData);
+                break;
+            case 'string':
+                let buf = new Buffer.from(data);
+                transData = buf.toString('utf8')
+                callback(transData);
+                break;
+            case 'boolean':
+                if (data[0] == 0 || data[0] == 1){
+                    transData = Boolean(data[0]);
+                } else {
+                    transData = null;
+                }
+                callback(transData);
+                break;
+            default:
+                logger.error('unknown dataType: ', property.datatype);
+                callback(null);
+                break;    
+        }
+    }
+
+    // updateActual update actual value to edge mqtt
+    updateActual(deviceID, property, value) {
```

Comment:
just fyi , we have changed from "expected/actual" to "desired/reported" in CRDs.

Reply:
this is still not reflected on edge side , overall looks inconsistent from UX.

Reply:
The edge side devicetwin is still using the "expected/actual", so just keep this format